### ISSUE TITLE
Support returning engines to pool via scope/using

### ISSUE THREAD

**Issue by binki** [Daniel15/JSPool#7]
I’m looking at the pattern suggested in the README:

```
var engine = pool.GetEngine();
var message = engine.CallFunction<string>("sayHello", "Daniel");
Console.WriteLine(message); // "Hello Daniel!"

// Always release an engine when you're done with it. This adds the engine back
// into the pool so it can be used again.
pool.ReturnEngineToPool(engine);
```

I think it would be nicer to be able to do something like:

```
using (var scopedEngine = pool.GetScopedEngine())
{
    // Access the JavaScriptEngineSwitcher interface either by
    // the Engine property or store it in a local variable for less typing:
    var engine = scopedEngine.Engine;
    var message = engine.CallFunction<string>("sayHello", "Daniel");

    // Let scoping call Dispose() for us and take care
    // of returning the engine to the pool. No more accidents!
}
```

Though it would be more fragile (if JavaScripEngineSwitcher adds more methods to `IJsEngine`, we’d break until fixed), it might even be more convenient if the object returned by `pool.GetScopedEngine()` implemented `IJsEngine` and proxied everything but `Dispose()` down to the actual `IJsEngine`. Then it could be even more sugary:

```
using (var engine = pool.GetScopedEngine())
{
    var message = engine.CallFunction<string>("sayHello", "Daniel");

    // Let scoping call Dispose() for us and take care
    // of returning the engine to the pool. No more accidents!
}
```

I have chosen a new function name, `IJsPool.GetScopedEngine()`, because existing code written against `IJsPool.GetEngine()` might make assumptions that this change in API would render invalid. Or, perhaps you’re willing to make a breaking change since this library is still a bit younger and `using(){}` is awesome enough to be worth it? If the scoping wrapper implemented `IJsEngine` and proxied methods to the wrapped real `IJsEngine`, the change in API would be transparent to the caller (you could leave `ReturnEngineToPool()` around and mark it with [ObsoleteAttribute](https://msdn.microsoft.com/en-us/library/system.obsoleteattribute%28v=vs.110%29.aspx) to encourage adoption of `using(){}`).

Thoughts? For now, I’m implementing something like this in my private code (using extension methods), but being able to use `using(){}` out of the box would make this library feel more complete.

**Comment by Daniel15** [comment]
Yeah, that's a good idea! 

My original use case was having one engine per web request so I couldn't use `using`, as the place I get the engine for the current request is different to the place I dispose the engine, it's not a single code block. Having said that, making `Dispose` return the engine to the pool would probably actually make it easier to use this library with dependency injection, as IoC containers call `Dispose` when stuff falls out of scope (eg. for per-request singletons once the request is finished).

I think the issue will be that calling `Dispose` today will actually dispose the engine itself. I'd need to have a proxy class that wraps the real engine, keeps track of which pool it belongs to, and overrides `Dispose`. I already do this for the MSIE engine since it can only be used from the same thread it was created on ([JsEngineWithOwnThread](https://github.com/Daniel15/JSPool/blob/master/src/JSPool/JsEngineWithOwnThread.cs)), but this change would involve doing it for all engines. That might be fine though.

> I have chosen a new function name, IJsPool.GetScopedEngine(), because existing code written against IJsPool.GetEngine() might make assumptions that this change in API would render invalid.

I think it'd be fine to just use `GetEngine` . This change would only change what `Dispose` does, and code shouldn't be touching Dispose right now (engines should always be returned to the pool).

**Comment by Daniel15** [comment]
Released in 3.0
